### PR TITLE
Added `Document` resource support

### PR DESCRIPTION
**Pull request by rushmash91**
Issue #, if available:

Description of changes:
Added support for Document along with e2e tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: apis/v1alpha1/ack-generate-metadata.yaml [hunk 1]
```diff
@@ -1,13 +1,13 @@
 ack_generate_info:
-  build_date: "2024-10-10T03:56:07Z"
-  build_hash: 36c2d234498c2bc4f60773ab8df632af4067f43b
+  build_date: "2024-11-09T00:29:41Z"
+  build_hash: 9715a2a715317a76ae83825294ca50cde9afd97b
   go_version: go1.23.2
-  version: v0.39.1
-api_directory_checksum: 8882bcc0e344c5efe175e36d5a8e75bcaca88174
+  version: 9715a2a
```
Comment (a-hilaly) [inline apis/v1alpha1/ack-generate-metadata.yaml:5]: Can you please pull the code-gen tags (`git pull --tags --upstream`) and regenerate the controller?

File: apis/v1alpha1/generator.yaml [hunk 2]
```diff
@@ -3,8 +3,15 @@ ignore:
...
       - OpsItem
       - PatchBaseline
       - ResourceDataSync
+
+resources:
+  Document:
+    exceptions:
+      errors:
+        404:
+          code: InvalidDocument
```
Comment (a-hilaly) [inline apis/v1alpha1/generator.yaml:17]: Does the API return 1InvalidDocument` when a document doesn't exist?

File: apis/v1alpha1/types.go [hunk 3]
```diff
@@ -27,3 +27,568 @@ var (
...
+	Values []*string `json:"values,omitempty"`
+}
+
+// Parameters specified in a Systems Manager document that run on the server
+// when the command is run.
+type DocumentParameter struct {
+	DefaultValue *string `json:"defaultValue,omitempty"`
+	Description  *string `json:"description,omitempty"`
+	Name         *string `json:"name,omitempty"`
+	Type         *string `json:"type_,omitempty"`
```
Comment (a-hilaly) [inline apis/v1alpha1/types.go:320]: Can we add a config to rename this field?
```suggestion
	Type         *string `json:"type,omitempty"`
```

here is an example in eks controller https://github.com/aws-controllers-k8s/eks-controller/blob/main/generator.yaml#L483-L484

File: metadata.yaml
Comment (a-hilaly) [inline metadata.yaml:1]: 👍

File: pkg/resource/document/sdk.go [hunk 4]
```diff
@@ -0,0 +1,1034 @@
...
+		ko.Status.StatusInformation = resp.StatusInformation
+	} else {
+		ko.Status.StatusInformation = nil
+	}
+	if resp.VersionName != nil {
+		ko.Spec.VersionName = resp.VersionName
+	} else {
+		ko.Spec.VersionName = nil
+	}
+
```
Comment (a-hilaly) [inline pkg/resource/document/sdk.go:170]: it looks like [SSM Document have a state](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_DocumentDescription.html#systemsmanager-Type-DocumentDescription-Status), so probably we want to requeue in both sdkFind and sdkCreate until the Docment is in ACTIVE state.

File: pkg/resource/document/sdk.go [hunk 5]
```diff
@@ -0,0 +1,1034 @@
...
+				f27elem.Key = f27iter.Key
+			}
+			if f27iter.Value != nil {
+				f27elem.Value = f27iter.Value
+			}
+			f27 = append(f27, f27elem)
+		}
+		ko.Spec.Tags = f27
+	} else {
+		ko.Spec.Tags = nil
```
Comment (a-hilaly) [inline pkg/resource/document/sdk.go:818]: I'm so glad the Update api supports tag updates too :)

**Review by a-hilaly — COMMENTED** [review]
Nice! Great job @rushmash91 ! I left a few comments/questions below

File: apis/v1alpha1/generator.yaml (hunk 2 above)
Comment (rushmash91) [reply apis/v1alpha1/generator.yaml:17]: Yes!

**Review by a-hilaly — APPROVED** [review]
Great work on this @rushmash91 🚀 
/lgtm

**Pull request merged**